### PR TITLE
adds db calls for subscriptions

### PR DESCRIPTION
**Pull request by tonytheleg**
### What this PR does
**Before this PR:**
* Subscription data was only stored in cache and not cosmos

**After this PR:**
* Subscription data is put into Cosmos DB
* Subscription data can be fetched and updated
* Document types are combined into single `documents.go`

Jira: [ARO-7410](https://issues.redhat.com/browse/ARO-7410)
Link to demo recording: 

### Special notes for your reviewer

Validation of all changes available in Jira comments ([LINK](https://issues.redhat.com/browse/ARO-7410?focusedId=24786971&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-24786971)

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Deployment: The deployment process was considered and addressed if required
- [ ] Testing: New code requires new unit tests.
- [ ] Documentation: Is the documentation updated? Either in the doc located in focus area, in the README or in the code itself.
- [ ] Customers: Is this change affecting customers? Is the release plan considered?

### CODE REVIEW

File: frontend/database.go
Comment (mjlshen) [inline frontend/database.go:192]: Thinking out loud, for this and `GetClusterDoc` should "not found" always return an error? Do we need to boolean return to identify if an object was found or not?

**Review by mjlshen — APPROVED** [review]
/lgtm, non-blocking question

File: frontend/database.go
Comment (mjlshen) [reply frontend/database.go:192]: Oh, yeah, we could do https://go.dev/blog/go1.13-errors instead of string comparison if we wanted to get rid of the boolean return value.

Comment (tonytheleg) [reply frontend/database.go:192]: Sounds like a plan, ill work on getting this updated to define an error and check the error type instead.

Comment (tonytheleg) [reply frontend/database.go:195]: @mjlshen updated, lemme know what you think, its better IMO. Cleaner.

File: frontend/frontend.go
Comment (mbarnes) [inline frontend/frontend.go:322]: This check needs to come before `if err != nil` or else we'll end up treating `ErrNotFound` like any other error.

File: frontend/database.go
Comment (mbarnes) [reply frontend/database.go:195]: +1 on eliminating the boolean, but as I noted elsewhere be careful in error handling to check specifically for `NotFoundError` _before_ `err != nil`.

File: frontend/frontend.go
Comment (tonytheleg) [reply frontend/frontend.go:322]: good catch, thanks! Change coming

Comment (tonytheleg) [reply frontend/frontend.go:317]: ok should be updated everywhere to 
1) Check for the ErrNotFound first when it make sense to
2) Finishes implementing the change from bool to error for Subscription docs too

**Comment by tonytheleg** [comment]
Throwing this in draft as there is one other bit of logic i think needs updating and want to make sure its all tested.

**Review by mjlshen — APPROVED** [review]

File: frontend/frontend.go
Comment (mbarnes) [inline frontend/frontend.go:396]: Quoth the RP Contract for resource deletion:
> The resource provider can return 200 (OK) or 204 (NoContent) to indicate that the operation completed successfully. A 200 (OK) should be returned if the object exists and was deleted successfully; and a 204 (NoContent) should be used if the resource does not exist and the request is well formed.

So I think we want `http.StatusNoContent` here.

File: frontend/frontend.go
Comment (mbarnes) [inline frontend/frontend.go:478]: This error handling logic flow seems a little clunky to me.  I get that you're trying to handle actual error conditions first, but I think it would read better if you structured it as you've done elsewhere:

```
if err != nil {
        if errors.Is(err, ErrNotFound) {
                // initialize new SubscriptionDocument
        } else {
                // write error and return
        }
} else {
        // use found SubscriptionDocument
}
```

Just my opinion though.  If you disagree then disregard.

**Review by mbarnes — CHANGES_REQUESTED** [review]
Couple more nits.  Requested change is for the status code, the other comment is optional.

File: frontend/frontend.go
Comment (tonytheleg) [reply frontend/frontend.go:396]: Good ole RP requirements haha. Easy enough fix though, thanks for catching it!

File: frontend/frontend.go
Comment (tonytheleg) [reply frontend/frontend.go:478]: Don't have strong opinions here, your ask makes sense, happy to add the change while im making the other one.

Comment (tonytheleg) [reply frontend/frontend.go:478]: I remember what my thinking was here. Since this call is being used to create or update, i was trying to avoid setting the `doc.Subscription` value before checking for all possible errors. I think what you've suggested does the same so that works for me, i think i just missed this way of getting to the same end. Thanks!

**Comment by tonytheleg** [comment]
Requested changes added

**Review by mbarnes — APPROVED** [review]

**Review by mbarnes — DISMISSED** [review]
Approved once the merge conflict is resolved.

**Review by mbarnes — APPROVED** [review]

**Pull request merged**
